mailpoet/lib/Validator/Validator.php

```diff
@@ -0,0 +1,190 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Validator;
+
+use JsonSerializable;
+use MailPoet\WP\Functions as WPFunctions;
+use stdClass;
+use WP_Error;
+
+class Validator {
+  /** @var WPFunctions */
+  private $wp;
+
+  public function __construct(
+    WPFunctions $wp
+  ) {
+    $this->wp = $wp;
+  }
+
+  /**
+   * Strict validation & sanitization implementation.
+   * It only coerces int to float (e.g. 5 to 5.0).
+   *
+   * @param mixed $value
+   * @return mixed
+   */
+  public function validate(Schema $schema, $value, string $paramName = 'value') {
+    $result = $this->validateAndSanitizeValueFromSchema($value, $schema->toArray(), $paramName);
+    if ($result instanceof WP_Error) {
+      throw ValidationException::createFromWpError($result);
+    }
+    return $result;
```

Comment:
This is a really basic and maybe silly question that probably demonstrates that I know nothing about what we're trying to achieve here, but why do we want to have validation and sanitization combined like this? Why not have 

```php
/**
 * @param mixed $value
 * @throws ValidationException
 */
public function validate(Schema $schema, $value, string $paramName = 'value'): void {
  $result = $this->validateAndSanitizeValueFromSchema($value, $schema->toArray(), $paramName);
  if ($result instanceof WP_Error) {
    throw ValidationException::createFromWpError($result);
  }
}
```

and then continue to use the original value if validation doesn't throw an exception?

It seems like if it passes validation it should be fine to use as-is, and then we don't have to worry about the order of schemas in `anyOf` definitions.

I suppose we always have the option to ignore the return value and continue using the original if we feel like it. I just want to make sure I understand why it's assumed we would always use the sanitized version.

Reply:
Using the code snippet you've shared it would be impossible to get the sanitized version. In the current form, you're free to choose whether you use the sanitized version or the original one.

But given the discussed issues with integers and floats and the fact that WP performs further sanitization (e.g. a string of type URL gets normalized to URL-encoded version), it's always strongly recommended to use the sanitized data. As we can [find in the WP docs](https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#api):

> When calling these functions, you should take care to always first validate the data using rest_validate_value_from_schema, and then if that function returns true, sanitize the data using rest_sanitize_value_from_schema. Not using both can open up your endpoint to security vulnerabilities.